src/uct/tcp/tcpcm/tcpcm_ep.c

```diff
@@ -0,0 +1,298 @@
+        } \
+    } while (0)
+
+ucs_status_t uct_tcpcm_ep_set_sock_id(uct_tcpcm_iface_t *iface, uct_tcpcm_ep_t *ep)
+{
+    ucs_status_t status;
+    struct sockaddr *dest_addr = NULL;
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+
+    /* create a sock_id for the client side */
+    if (iface->sock_id_quota > 0) {
+        /* Create an id for this interface. Events associated with this id will be
+         * reported on the event_channel that was created on iface init. */
+        ep->sock_id_ctx = ucs_malloc(sizeof(*ep->sock_id_ctx), "client sock_id_ctx");
+        if (ep->sock_id_ctx == NULL) {
+            status = UCS_ERR_NO_MEMORY;
+            goto out;
+        }
+
+        /* FIXME: Review */
+        dest_addr = (struct sockaddr *) &(ep->remote_addr);
+
+        status = ucs_socket_create(dest_addr->sa_family, SOCK_STREAM,
+                                   &ep->sock_id_ctx->sock_id);
+        if (status != UCS_OK) {
+            goto out_free;
+        }
+
+        ep->sock_id_ctx->ep = ep;
+        ucs_list_add_tail(&iface->used_sock_ids_list, &ep->sock_id_ctx->list);
+        iface->sock_id_quota--;
+        ucs_debug("ep %p, new sock_id %d. sock_id_in_quota %d", ep,
+                   ep->sock_id_ctx->sock_id, iface->sock_id_quota);
+        status = UCS_OK;
+        goto out;
+    } else {
+        ep->sock_id_ctx = NULL;
+        status = UCS_ERR_NO_RESOURCE;
+        goto out;
+    }
+
+out_free:
+    ucs_free(ep->sock_id_ctx);
+out:
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+    return status;
+}
+
+ucs_status_t uct_tcpcm_send_client_info(uct_tcpcm_iface_t *iface, uct_tcpcm_ep_t *ep)
+{
+    uct_tcpcm_conn_param_t conn_param;
+    uct_tcpcm_priv_data_hdr_t *hdr;
+    ssize_t sent_len = 0;
+    ssize_t recv_len = 0;
+    ssize_t offset = 0;
+    int connect_confirm = -1;
+
+    memset(&conn_param.private_data, 0, UCT_TCPCM_UDP_PRIV_DATA_LEN);
+    hdr = &conn_param.hdr;
+
+    /* pack worker address into private data */
+    hdr->length = ep->pack_cb(ep->pack_cb_arg, "eth0",
+                              (void*)conn_param.private_data);
+    if (hdr->length < 0) {
+        ucs_error("tcpcm client (iface=%p) failed to fill "
+                  "private data. status: %s",
+                  iface, ucs_status_string(hdr->length));
+        return UCS_ERR_IO_ERROR;
+    }
+    hdr->status = UCS_OK;
+    conn_param.private_data_len = sizeof(uct_tcpcm_conn_param_t);
+
+    // send all of the connection info
+
+    //if (UCS_OK != ucs_sys_fcntl_modfl(ep->sock_id_ctx->sock_id, O_NONBLOCK, 0)) {
+    //    ucs_error("fcntl() failed: %m");
+    //    return UCS_ERR_IO_ERROR;
+    //}
+
+    recv_len = recv(ep->sock_id_ctx->sock_id, (char *) &connect_confirm,
+                    sizeof(int), 0);
+    ucs_debug("recv len = %d\n", (int) recv_len);
+    ucs_debug("connect confirm = %d\n", connect_confirm);
+
+    sent_len = send(ep->sock_id_ctx->sock_id, (char *) &conn_param + offset,
```

Comment:
UCT/TCP has non-blocking variant for send/recv.
I guess we can move it to the common code and implement the blocking variants for send/recv on top of the non-blocking ones (it will work for both blocking and non-blcoking socket fds)

Reply:
@Akshay-Venkatesh I've added `ucs_socket_send`/`ucs_socket_recv` and `ucs_socket_send_nb`/`ucs_socket_send_nb` in the `ucs/sys/sock.[ch]` to use from the UCT/TCP and UCT/TCPCM code. Please, take a look at the PR #3423

Reply:
btw, the PR #3431 has been merged. you can use `ucs_socket_send`/`ucs_socket_recv` (blocking variants) here even if your socket fd is marked as non-blocking one.

Reply:
Will do.